### PR TITLE
Implemented partial sum cache for SMAIndicator

### LINKED ISSUE #1139: SMAIndicator inner cache
**I have checked existing issues and wiki**
- [x] I could not find similar [issues](https://github.com/ta4j/ta4j/issues?utf8=%E2%9C%93&q=)
- [x] I could not find a solution in the [wiki](https://ta4j.github.io/ta4j-wiki/) or [faq section](https://ta4j.github.io/ta4j-wiki/FAQ.html)

**Is your feature request related to a problem? Please describe.**
SMAIndicator may benefit from reusage of precalculated partial sums.
CachedIndicator caches only final result, but, for example, SMA 200 have to iterate over 200 bars for each new bar again and again.
We can store for serial access previous sum and then subtract FIFO bar which gous out and add a new bar which goes in.

### PR DESCRIPTION
**Pull request by sgflt**
Fixes #1139.

Performance of partial sums.

Changes proposed in this pull request:
- longer barSeries benefits from reusage of precomputed partial sums
- random access fallbacks to slowPath

- [x] added an entry with related ticket number(s) to the unreleased section of `CHANGES.md`

### CODE REVIEW

**Comment by sgflt** [comment]
Test on real data:
**Before:**
ta4jexamples.backtesting.MovingAverageCrossOverRangeBacktest -- Back-tested 2145 strategies on 6150-bar series in 72520 ms


**After:**
ta4jexamples.backtesting.MovingAverageCrossOverRangeBacktest -- Back-tested 2145 strategies on 6150-bar series in 6222 ms

File: ta4j-core/src/main/java/org/ta4j/core/indicators/SMAIndicator.java [hunk 1]
```diff
@@ -36,6 +36,11 @@ public class SMAIndicator extends CachedIndicator<Num> {
 
     private final Indicator<Num> indicator;
     private final int barCount;
+    private Num previsouSum = zero();
```
Comment (TheCookieLab) [inline ta4j-core/src/main/java/org/ta4j/core/indicators/SMAIndicator.java]:
```suggestion
    private Num previousSum = zero();
```

**Review by TheCookieLab — CHANGES_REQUESTED** [review]

**Comment by sgflt** [comment]
I will add edge case tests, to prove functionality.

**Comment by sgflt** [comment]
Umm, what is expected behavior for moving series?

If I add following test:

```java
    @Test
    public void usingBarCount3UsingClosePriceMovingSerie() {
        data.setMaximumBarCount(13);
        data.addBar(new MockBar(5., numFunction));

        Indicator<Num> indicator = getIndicator(new ClosePriceIndicator(data), 3);

        assertNumEquals((0d + 0d + 2d) / 1, indicator.getValue(data.getBeginIndex() + 0));
        assertNumEquals((0d + 2d + 3d) / 2, indicator.getValue(data.getBeginIndex() + 1));
        assertNumEquals((2d + 3d + 4d) / 3, indicator.getValue(data.getBeginIndex() + 2));
        assertNumEquals((3d + 4d + 3d) / 3, indicator.getValue(data.getBeginIndex() + 3));
    }

```


It fails on second assert (0d + 2d + 3d) / 2 even with previous implementation.

```
Expected :2.5
Actual   :2.3333333333333335
```

**Comment by sgflt** [comment]
Discovered RunningTotalIndicator. I may use this indicator as partialSum and optimize RunningTotalIndicator.

**Comment by sgflt** [comment]
And now I have discovered PreviousValueIndicator. I think I may combine them too.

**Comment by sgflt** [comment]
Nope with indexes.

**Comment by TheCookieLab** [comment]
Yes, you'll want to handle unstable indexes as unstable and assert accordingly.

**Comment by TheCookieLab** [comment]
@sgflt please run the formatter (`mvn -B clean license:format formatter:format test`) and re-push

**Review by TheCookieLab — APPROVED** [review]

**Comment by TheCookieLab** [comment]
@sgflt I recently cut the 0.16 release and started 0.17. Can you update your CHANGELOG entry accordingly? Thanks.

**Comment by sgflt** [comment]
Done

**Review by TheCookieLab — APPROVED** [review]
